components/SerialServer/src/serial.c

```diff
@@ -310,44 +268,89 @@ static void internal_putchar(int b, int c)
     error = serial_unlock();
 }
 
-static void internal_raw_putchar(int id, int c)
+static void serial_notify_free_send(virtqueue_driver_t *queue)
 {
-    getchar_client_t *client = &getchar_clients[id];
-    uint32_t next_tail = (client->buf->tail + 1) % sizeof(client->buf->buf);
-    if (next_tail == client->buf->head) {
-        /* full */
-        return;
+    void *buf = NULL;
+    size_t buf_size = 0, wr_len = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+    while (virtqueue_get_used_buf(queue, &handle, &wr_len)) {
+        while (camkes_virtqueue_driver_gather_buffer(queue, &handle, &buf, &buf_size, &flag) >= 0) {
+            /* Clean up and free the buffer we allocated */
+            camkes_virtqueue_buffer_free(queue, buf);
+        }
     }
-    uint32_t last_read_head = client->buf->head;
-    client->buf->buf[client->buf->tail] = (uint8_t)c;
-    /* no synchronize in here as we assume TSO */
-    client->buf->tail = next_tail;
-    __sync_synchronize();
-    if (last_read_head != client->last_head) {
-        getchar_emit(client->client_id);
-        client->last_head = last_read_head;
+}
+
+static int serial_notify_recv(int client, virtqueue_device_t *queue)
+{
+    int err;
+    void *buf = NULL;
+    size_t buf_size = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_available_buf(queue, &handle)) {
+        char temp_buf[BUFSIZE];
```

Comment:
I wonder if this buffer might better come from a client context than on the stack? That make the resource usage a bit more transparent and put less load on the stack.